### PR TITLE
Contribution and PR Guidelines

### LINKED ISSUE #160: Branch protection, Contributing guidelines, PR templates?
## Proposed branch protection for `master`:
image

## Contributing guidelines
- Put in `CONTRIBUTING.md`
- instructions for 3rd party contributors
- mention in README

## PR templates
- include a checklist of things for PRs
- Put in `/.github/pull_request_template.md`

### PR DESCRIPTION
**Pull request by classabbyamp**
Closes #160

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+### Checklist
+
+- [ ] Code reviewed by the author
+- [ ] Code documented (comments or other documentation)
+- [ ] Changes tested
+- [ ] `flake8` passes
+- [ ] `CHANGELOG.md` updated
+- [ ] Informative commit messages
+- [ ] Issue exists for PR
+- [ ] Descriptive PR title
```
[the remark this answers is not in the document]
Comment (0x5c) [reply .github/pull_request_template.md:29]: I would make "Issue exists for PR" as the first element

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+### Environment Setup
+
+Once all of the above is done, you can get started by setting up your development envronment.
+
+1. [Fork this repo][1] into your own GitHub namespace.
+1. Make sure the `master` branch is up to date, then make yourself a new branch with a descriptive name.
+1. Once the forked repo is cloned and on the proper branch, you can set up the development environment.
+    1. Run `make dev-install`.
+       This should install everything you need to develop and run qrm.
```
Comment (0x5c) [inline CONTRIBUTING.md:18]: Not true, python3.7+ is required, and 3.8+ if we decide to bump the version requirement up.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    1. [Create a bot and token][2], and add it to `data/keys.py`.
+       Also add your [QRZ credentials][3] if needed.
+    1. In `data/options.py`, change values as needed.
+       Some commands require adding your Discord user ID to `owner_uids`.
+    1. To activate the virtual env that was created by `make`, run `source botenv/bin/activate` (or the equivelent for your shell or operating system).
+
+## While You Develop
+
+To run qrm, use the command `sh run.sh`.
```
Comment (0x5c) [inline CONTRIBUTING.md:27]: `./run.sh`

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+If you want to break when an error happens, use the flag `--pass-errors`.
```
Comment (0x5c) [inline CONTRIBUTING.md:28]:
```suggestion
We recommend you use the `--pass-errors` flags to avoid perpetual restart loops if you break the bot. It exists because repeatedly mashing [Ctrl+C] at high speed to break a fast loop is not fun.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+Make commits as needed, but try to keep it reasonable.
+If there are too many, your contribution may be squashed when merged.
+Commit messages should be descriptive and mention issues that they fix ("fixes #123") or contain progress on ("progress on #123").
+
+Make sure to document your code as you go, in both comments and external documentation (in `/dev-notes/`) as needed.
```
Comment (0x5c) [inline CONTRIBUTING.md:34]:
```suggestion
Make sure to document your code as you go, in both comments and external documentation (in `/dev-notes/`) as needed. `dev-notes` is especially important if you introduce a new json file format or to document some development process (like the command to crush the various images in the repository).
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+       Some commands require adding your Discord user ID to `owner_uids`.
+    1. To activate the virtual env that was created by `make`, run `source botenv/bin/activate` (or the equivelent for your shell or operating system).
+
+## While You Develop
+
+To run qrm, use the command `sh run.sh`.
+If you want to break when an error happens, use the flag `--pass-errors`.
+
+Make commits as needed, but try to keep it reasonable.
+If there are too many, your contribution may be squashed when merged.
```
Comment (0x5c) [inline CONTRIBUTING.md:31]: We need to recommend to recommit when the PR is ready.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+Commit messages should be descriptive and mention issues that they fix ("fixes #123") or contain progress on ("progress on #123").
+
+Make sure to document your code as you go, in both comments and external documentation (in `/dev-notes/`) as needed.
+
+**Test your changes.**
+If your code doesn't work, it's not ready for merging.
+Make sure you not only test intended behaviour, but also edge cases and error cases.
```
Comment (0x5c) [inline CONTRIBUTING.md:38]: Add flake8 and mypy.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+if that fails, you will be expected to fix those errors before merge.
+
+Otherwise, try to follow the existing style:
+double-quotes except when required to be single,
+indentation of mult-line structures matching other examples in the code,
+etc.
+
+## When You're Ready to Merge
+
+1. When you have finished working on your contribution, create a pull request from your fork's branch into the main master branch.
```
Comment (0x5c) [inline CONTRIBUTING.md:57]:
```suggestion
1. When you have finished working on your contribution, create a pull request from your fork's branch into the master branch of this repository.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+1. Read through and complete the pull request template.
+   If the checklist is not complete, your contribution will not be merged.
+1. Your pull request will get reviewed by at least one maintainer.
+1. If approved, another maintainer will merge the pull request.
```
Comment (0x5c) [inline CONTRIBUTING.md:61]: Merge if no additional requests

**Review by 0x5c — CHANGES_REQUESTED** [review]
A couple of small changes

File: .github/pull_request_template.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+Fixes #{issue}
+
+### Type of change
+
+Please delete options that are not relevant.
+
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
+- [ ] This change requires a documentation update
```
[the remark this answers is not in the document]
Comment (classabbyamp) [reply .github/pull_request_template.md:14]: yes I noticed that, I think the way to do it is to remove the line about deleting things.

File: CONTRIBUTING.md (hunk 2 above)
Comment (classabbyamp) [reply CONTRIBUTING.md:18]: ah yes true

File: CONTRIBUTING.md (hunk 7 above)
Comment (classabbyamp) [reply CONTRIBUTING.md:38]: I believe I mentioned flake8 elsewhere but I will add mypy.

File: .github/pull_request_template.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (0x5c) [reply .github/pull_request_template.md:12]: Since those are simply flags, we need to not treat them as todo-list items.
So the line about deleting is ok, but the checkboxes not.

**Review by 0x5c — APPROVED** [review]
